Recognise new IPv6 non-global range from IETF RFC 9602

bardiharborow opened the pull request:
This PR adds the `5f00::/16` range defined by [IETF RFC 9602](https://datatracker.ietf.org/doc/rfc9602/) to those ranges which `Ipv6Addr::is_global` recognises as a non-global IP. This range is used for Segment Routing (SRv6) SIDs.

See also: https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml
Unstable tracking issue: #27709

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12856960436/job/35844267950) [(plain)](https://github.com/rust-lang/rust/commit/3807f524d87f32b492133b8d63f799d064c607db/checks/35844267950/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- net::ip_addr::ipv6_properties stdout ----

thread 'net::ip_addr::ipv6_properties' panicked at library/core/tests/net/ip_addr.rs:692:5:
assertion failed: !ip!("5f00::").is_unicast_global()

failures:
    net::ip_addr::ipv6_properties

```

</details>

bardiharborow commented on library/core/src/net/ip_addr.rs:
```diff
@@ -1773,6 +1775,8 @@ impl Ipv6Addr {
             && !self.is_documentation()
             && !self.is_benchmarking()
+            // Segment Routing (SRv6) SIDs (`5f00::/16`)
+            && !matches!(self.segments(), [0x5f00, ..])
```
🐃 I haven't changed this to something potentially more comprehensive like `self.is_unicast() && !self.global()` since that would change behaviour for some other ranges, but it may be a good target for another PR.

cuviper replied:
See #85604 -- I think given those open questions, we should not change this at all right now.

(The last `is_global` update in #119006 also avoided this.)

bardiharborow replied:
I will have a look and re-request review when this is ready.

@rustbot author

bardiharborow commented:
@rustbot label A-io

joboet commented:
This isn't my area of expertise, sorry...
r? libs

bors commented:
:umbrella: The latest upstream changes (presumably #135937) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

LFS6502 commented:
@bardiharborow 
Thanks for your contribution.
From wg-triage. Your PR now has merge conflicts, are you able to rebase on top of master and fix them? Thanks!

Dylan-DPC commented:
@bardiharborow any updates on this? thanks

bardiharborow committed:
Recognise new IPv6 non-global range from RFC9602

This commit adds the 5f00::/16 range defined by RFC9602 to those ranges which Ipv6Addr::is_global recognises as a non-global IP. This range is used for Segment Routing (SRv6) SIDs.

bardiharborow commented:
I've removed the part that was objectionable.

@rustbot ready

cuviper commented:
Thanks!

@bors r+ rollup

bors commented:
:pushpin: Commit ea99e81485ff5d82cabba9af5d1c21293737cc16 has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper ea99e81485ff5d82cabba9af5d1c21293737cc16 -->
<!-- homu: {"type":"Approved","sha":"ea99e81485ff5d82cabba9af5d1c21293737cc16","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
